### ISSUE TITLE
Add scrolling through playlist view while holding middle mouse button

### ISSUE THREAD

**Issue by Tenrys** [reupen/columns_ui#122]
I just switched to using this plugin in order to be able to use [this](https://github.com/Chocobo1/foo_uie_output_switcher/releases) and I just noticed I can't hold the middle mouse button to scroll the playlist view horizontally or vertically like I can with the default UI. 

How hard would it be to implement this feature?

**Comment by Tenrys** [comment]
Neato, I wasn't aware. Thanks for the tip, I'm using that now.

**Comment by MG240** [comment]
+1

**Comment by reupen** [comment]
Use the thumbs up reaction on the first comment as I can [easily sort by that](https://github.com/reupen/columns_ui/issues?q=is%3Aissue%20state%3Aopen%20sort%3Areactions-%2B1-desc) 🙂

**Comment by MG240** [comment]
Got it! Thank you

**Comment by MG240** [comment]
@reupen 

Thank you for the amazing enhancements and features you've put into Columns UI.

I'm curious about your method for scrolling long playlists. Do you depend entirely on the mouse wheel, drag the scrollbar thumb, or use the PgUp and PgDn buttons? Would an autoscroll implementation affect the performance of the recently implemented smooth scroll?

I'm asking since I'm used to autoscroll in web browsers and other applications, and I was wondering if there's a technical reason or argument against it.

**Comment by reupen** [comment]
Thanks!

I have a discontinued Microsoft Precision Mouse. It has a mouse wheel that supports proper free scrolling. So it scrolls effectively at pixel-level, rather than the fixed number of lines configured in Windows settings (that instead just affects how fast it scrolls). It's more like how scrolling with a modern touch pad is, and so I can scroll both quite fast and accurately easily. Additionally, I will frequently type into the playlist to jump to a particular artist (if the playlist is sorted by artist).

Unfortunately for me, that mouse is discontinued and there are few current mice that behave the same way, so when it breaks I might be buggered. (Logitech mice do have a smooth scrolling option, but it only works when specific apps like web browsers are focused so is useless for me, and it doesn't quite behave the same either IIRC).

There is no technical reason blocking implementing autoscroll. It's simply that I haven't had a need for it and it was complicated enough to implement that it hasn't been done yet. But it was on my mind already as something to finally get done so I may look at it soon.

**Comment by reupen** [comment]
There’s a version of this available to test attached to https://github.com/reupen/columns_ui/actions/runs/28626145671. (See also #1786.)

Like Chrome, Firefox and Notepad it will activate it on middle-click and it will stay activated until you click again (or press Esc). (Which is different from Default UI and the first comment in this issue as in Default UI you have to keep the middle-click button held down.)

It behaves most similarly to Chrome. It doesn’t leave a persistent indicator on screen at the initial middle-click point at the moment, though Chrome doesn’t either so perhaps that’s not really needed?

**Comment by MG240** [comment]
@reupen 

Thank you very much. This is amazing.

I hadn't noticed this difference between Chrome & Firefox before. The current implementation (Chrome style) looks cleaner, but the persistent circle indicator in Firefox is slightly more user-friendly; it acts as a visual reference point for the initial click, showing how far the mouse pointer is from it and, therefore, how fast the scrolling is. However, if it is more complex to implement, the current implementation is already excellent.

I also noticed that a mouse click deactivates the autoscroll, but it also selects the item beneath the pointer. I propose changing this so that the click only deactivates the autoscroll, without selecting the item.

**Comment by reupen** [comment]
Thanks for testing!

> I hadn't noticed this difference between Chrome & Firefox before. The current implementation (Chrome style) looks cleaner, but the persistent circle indicator in Firefox is slightly more user-friendly; it acts as a visual reference point for the initial click, showing how far the mouse pointer is from it and, therefore, how fast the scrolling is. However, if it is more complex to implement, the current implementation is already excellent.

Yes, I agree that it would be better to have it really (it’s what I expected and was used to, I was surprised that Chrome doesn’t have it to be honest). It is a fair bit of work to draw that properly but it is doable so I should be able to add it in.

> I also noticed that a mouse click deactivates the autoscroll, but it also selects the item beneath the pointer. I propose changing this so that the click only deactivates the autoscroll, without selecting the item.

Agree, will change this.

**Comment by MG240** [comment]
@reupen 

I've tried the latest [test build](https://github.com/reupen/columns_ui/actions/runs/28687948862) and the new autoscroll indicator looks great! One thing I noticed: rapid, consecutive middle-clicks don't register the second click. For example, quickly toggling autoscroll on and off (or off and on) in fast succession doesn't work. There needs to be a brief pause between clicks for the second one to register.

**Comment by reupen** [comment]
Fixed, thanks! (It was simply because it became a double-middle-click and that wasn't handled.)

Now, double-left-clicking behaves oddly (which is because the handling of the first left down message is suppressed) so I need to work out what to do about that...

**Comment by MG240** [comment]
@reupen 

Yes, double-left-clicking now turns autoscroll off, but it plays the track that was selected before autoscroll was enabled.

Also, double-middle-clicking the same spot to toggle autoscroll seems fixed now. However, if you turn autoscroll on and then double-middle-click a different spot, the second click isn't registered unless there is a brief pause between the clicks.

**Comment by MG240** [comment]
@reupen 

Another autoscroll behavior I'm used to, but never really paid attention to before, that exists in both Chrome and Firefox, is that if I middle-click, hold, and release before moving the cursor (or within a certain tiny distance of the initial click), the auto-scroll stays active. However, if I release the click past that distance (or make an autoscroll movement), the autoscroll turns off.

**Comment by reupen** [comment]
Thanks, I wasn’t aware of that behaviour existing (but it explains the Default UI behaviour and the original description of this issue...)

That should be implemented now. So it should just be the double-left-click thing left to sort out.

**Comment by MG240** [comment]
@reupen 

Thank you very much.

This might be a minor subjective preference, but I think it would be better if the mouse pointer didn't change styles when moving up and down during an autoscroll, especially to the style when it's on the same level as the initial autoscroll indicator, as there would be two similar indicators on the screen (the pointer itself and the initial persistent autoscroll). The initial persistent autoscroll indicator already provides enough clarity.

**Comment by reupen** [comment]
I did question whether the initial pointer change is necessary. It’s tricky as there is no reference implementation, and every implementation is different 🙃

I think the closest thing to a current reference Microsoft implementation is Word, which looks like the below:

https://github.com/user-attachments/assets/eb34b2aa-8aaa-48fa-bf75-7bb98db70d0e

You could consider Internet Explorer as the true reference, which looks like this in a Windows 7 VM:

https://github.com/user-attachments/assets/84c1e8ea-4a0f-4ba4-b45e-defc07dc16cf

Putting the stylistic differences of the pointers and the click-point indicator aside, they seem to behave similarly to what’s been implemented.

I also think it all also provides useful feedback (e.g. when you are actually in the 'do nothing' area or when doing a four-way scroll which has eight directions including the diagonals).

I might try changing the style of the initial pointer to make it similar to the other ones (without the background) though.
